### PR TITLE
Feature/printscreen

### PR DESCRIPTION
**Pull request by andydotxyz**
Add print screen features
* PrintScreen button captures all displays
* Shift+PrintScreen captures the current window with border.

The images are generated in the windowmanager code but our UI overlay then presents it to the user and offers to save.

### CODE REVIEW

**Review by Jacalz — COMMENTED** [review]
This is a great feature. I do have some issues with Fynedesk crashing when I try to take a screenshot of GTK apps like Firefox and it seems like capturing GTK apps doesn't work in full screen mode either.
firefox-empty

**Comment by andydotxyz** [comment]
From your screenshot it's hard to say what should have been shown - was firefox full screen on the left window?

**Comment by Jacalz** [comment]
Yeah that is a very true point. It was not full-screen, I think, but I do know that I tested Nautilus (the standard gnome file-manager) and it did not show either. Please also notice that the borders on `wormhole-gui` are gone.

**Comment by andydotxyz** [comment]
@Jacalz thinking more about this the missing borders were fixed in one of the later commits in this PR - can you please double check that you were testing against the latest version of this PR?

It is possibly still an issue for other reasons, but one issue was resolved during development :)

**Comment by andydotxyz** [comment]
And to confirm, when your windows are missing they are leaving those white rectangles rather than being omitted completely?

**Comment by Jacalz** [comment]
@andydotxyz I updated to the latest version and it appears to still be the case that borders are not shown. They do appear as dark in the preview for the screenshot though. Perhaps they are transparent? Yup, double checked in GIMP and they are indeed transparent.
image

Yes, the gtk windows are replaced with the white rectangles. I have never seen any windows be left out entirely. Please see the attached screenshot below:
firefox-empty

Aspect ratio for the screenshot viewer seems to be wrong until it gets some sort of response (click, drag or something that triggers a refresh) and then it gets the right aspect ratio.
strange-aspect

I have also been seeing once in a while that the screenshot viewer will not show any picture after taking a screenshot. I cannot replicate it reliably but have seen it once or twice at least.

**Comment by andydotxyz** [comment]
Thanks. The aspect issues were resolved in Fyne earlier today, and should be OK now.
I will try to replicate the GTK issues later tonight

**Comment by andydotxyz** [comment]
Strange, here is nautilus at my end...
nautilus

**Comment by andydotxyz** [comment]
Firefox worked too...
firefox

**Comment by andydotxyz** [comment]
Perhaps the issue @Jacalz finds is due to screen depth not matching the colour depth on the borders?
Can you share anything about the colour confugration in your Xorg @Jacalz.

Have you been able to test this yet @okratitan - it's been sitting for a month now and I'd like to gauge whether I am lucky in this working for me, or unlucky that it doesn't work on some old configuration...

**Comment by Jacalz** [comment]
@andydotxyz Sorry, but I am not entireley sure where to look for that sort of colour information. Do you perhaps know where that information is usually stored?

**Comment by andydotxyz** [comment]
It might be determined at runtime - but some X summary printout or maybe xrandr could list it?

**Comment by Jacalz** [comment]
Looks like I still have the same issues with windows being blank (transparent). Also does look like it can't capture the second monitor at all after the multidesk fix.

**Comment by Jacalz** [comment]
fynedesk-printscreen
This is what I am seeing on the latest commit. Everything that isn't a fyne application is transparent and the second monitor is not captured at all.

File: internal/ui/screenshot.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	preview := canvas.NewImageFromImage(img)
+	preview.FillMode = canvas.ImageFillContain
+	w.SetContent(container.NewBorder(nil, buttons, nil, nil, preview))
+	w.Resize(fyne.NewSize(400, 250))
+	w.Show()
+}
+
+func saveImage(pix image.Image, w fyne.Window) {
+	d := dialog.NewFileSave(func(write fyne.URIWriteCloser, err error) {
+		if err != nil {
```
Comment (Jacalz) [inline internal/ui/screenshot.go:57]: You are not checking if `write == nil`. I was noticing that pressing save and then cancelling crashes the desktop.

File: internal/ui/screenshot.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+		layout.NewSpacer(),
+		widget.NewButton("Cancel", func() {
+			w.Close()
+		}),
+		save)
+
+	preview := canvas.NewImageFromImage(img)
+	preview.FillMode = canvas.ImageFillContain
+	w.SetContent(container.NewBorder(nil, buttons, nil, nil, preview))
+	w.Resize(fyne.NewSize(400, 250))
```
Comment (Jacalz) [inline internal/ui/screenshot.go:51]: Could we perhaps make the window slightly larger? The window is too small to make the save dialog big enough to use.

**Review by Jacalz — CHANGES_REQUESTED** [review]
Thanks for the latest fix. Screenshots are now working :1st_place_medal: 
Just two smaller issues, but otherwise approved :+1:

**Review by Jacalz — APPROVED** [review]
Looks good to me. Staticcheck is failing, so you might need to manually merge this.

**Pull request merged**